src/Storages/VirtualColumnUtils.cpp

```diff
@@ -96,30 +122,63 @@ void rewriteEntityInAst(ASTPtr ast, const String & column_name, const Field & va
     }
 }
 
-void filterBlockWithQuery(const ASTPtr & query, Block & block, const Context & context)
+bool prepareFilterBlockWithQuery(const ASTPtr & query, const Context & context, Block block, ASTPtr & expression_ast)
 {
+    bool unmodified = true;
     const auto & select = query->as<ASTSelectQuery &>();
     if (!select.where() && !select.prewhere())
-        return;
+        return unmodified;
 
-    NameSet columns;
-    for (const auto & it : block.getNamesAndTypesList())
-        columns.insert(it.name);
+    ASTPtr condition_ast;
+    if (select.prewhere() && select.where())
+        condition_ast = makeASTFunction("and", select.prewhere()->clone(), select.where()->clone());
+    else
+        condition_ast = select.prewhere() ? select.prewhere()->clone() : select.where()->clone();
+
+    // Prepare a constant block with valid expressions
+    for (size_t i = 0; i < block.columns(); ++i)
+        block.getByPosition(i).column = block.getByPosition(i).type->createColumnConstWithDefaultValue(1);
```

Comment:
It may generate invalid expression, so query will fail with exception. 
Example:
```
:) create table mt (n UInt64) engine=MergeTree order by n partition by n % 10
:) insert into mt select * from numbers(100)
:) select distinct _part, toUInt64(substr(_part, 1, position(_part, '_') - 1)) from mt format TSV -- works fine

9_10_10_0	9
6_7_7_0	6
2_3_3_0	2
5_6_6_0	5
4_5_5_0	4
3_4_4_0	3
8_9_9_0	8
7_8_8_0	7
1_2_2_0	1
0_1_1_0	0

:) select * from mt where toUInt64(substr(_part, 1, position(_part, '_') - 1)) = 1 -- fails because empty string (default value) makes the expression invalid

Received exception from server (version 21.5.1):
Code: 32. DB::Exception: Received from localhost:9000. DB::Exception: Attempt to read after eof: Cannot parse UInt64 from String, because value is too short: While processing toUInt64(substring(_part, 1, position(_part, '_') - 1)) = 1.
```

Another example: https://github.com/ClickHouse/ClickHouse/blob/ad414bab055713882cfa7e16af022b1dbde2c1bd/tests/queries/0_stateless/01151_storage_merge_filter_tables_by_virtual_column.sql#L15

Reply:
@amosbird, please take a look

Reply:
Sure.